Userland/Libraries/LibJS/CyclicModule.cpp

```diff
@@ -0,0 +1,658 @@
+
+    // 2. If module.[[Status]] is linking, linked, evaluating-async, or evaluated, then
+    if (m_status == ModuleStatus::Linking || m_status == ModuleStatus::Linked || m_status == ModuleStatus::EvaluatingAsync || m_status == ModuleStatus::Evaluated) {
+        // a. Return index.
+        return index;
+    }
+
+    // 3. Assert: module.[[Status]] is unlinked.
+    VERIFY(m_status == ModuleStatus::Unlinked);
+
+    // 4. Set module.[[Status]] to linking.
+    m_status = ModuleStatus::Linking;
+
+    // 5. Set module.[[DFSIndex]] to index.
+    m_dfs_index = index;
+
+    // 6. Set module.[[DFSAncestorIndex]] to index.
+    m_dfs_ancestor_index = index;
+
+    // 7. Set index to index + 1.
+    ++index;
+
+    // 8. Append module to stack.
+    stack.append(this);
+
+    dbgln_if(JS_MODULE_DEBUG, "[JS MODULE] module: {} has requested modules: [{}]", filename(), String::join(",", m_requested_modules));
+
+    // 9. For each String required of module.[[RequestedModules]], do
+    for (auto& required_string : m_requested_modules) {
+        ModuleRequest required { required_string };
+
+        // a. Let requiredModule be ? HostResolveImportedModule(module, required).
+        auto required_module = TRY(vm.host_resolve_imported_module(this, required));
+
+        // b. Set index to ? InnerModuleLinking(requiredModule, stack, index).
+        index = TRY(required_module->inner_module_linking(vm, stack, index));
+
+        // c. If requiredModule is a Cyclic Module Record, then
+        if (is<CyclicModule>(*required_module)) {
+            auto& cyclic_module = static_cast<CyclicModule&>(*required_module);
+            // i. Assert: requiredModule.[[Status]] is either linking, linked, evaluating-async, or evaluated.
+            VERIFY(cyclic_module.m_status == ModuleStatus::Linking || cyclic_module.m_status == ModuleStatus::Linked || cyclic_module.m_status == ModuleStatus::EvaluatingAsync || cyclic_module.m_status == ModuleStatus::Evaluated);
+
+            // ii. Assert: requiredModule.[[Status]] is linking if and only if requiredModule is in stack.
+            VERIFY((cyclic_module.m_status == ModuleStatus::Linking) == (stack.contains_slow(&cyclic_module)));
+
+            // iii. If requiredModule.[[Status]] is linking, then
+            if (cyclic_module.m_status == ModuleStatus::Linking) {
+                // 1. Set module.[[DFSAncestorIndex]] to min(module.[[DFSAncestorIndex]], requiredModule.[[DFSAncestorIndex]]).
+                m_dfs_ancestor_index = min(m_dfs_ancestor_index.value(), cyclic_module.m_dfs_ancestor_index.value());
+            }
+        }
+    }
+
+    // 10. Perform ? module.InitializeEnvironment().
+    (void)TRY(initialize_environment(vm));
+
+    // 11. Assert: module occurs exactly once in stack.
+    auto count = 0;
+    for (auto* module : stack) {
+        if (module == this)
+            count++;
+    }
+    VERIFY(count == 1);
```

Comment:
I notice this while watching @linusg's video.

Wouldn't it be more correct to use `size_t` here?

`stack`  is an `AK::Vector` that can have up to `SIZE_MAX` elements.
Here `count` would be a `signed int`.

Also the only purpose of this code is to assert that `count == 1` so it is probably not great that a `SIZE_MAX` (on 64 bit) can overflow a `signed int` 4_294_967_295 times (maybe my math is slightly wrong, but somewhere around that number :D).

Reply:
merged 13 seconds ago?? man, my timing sucks!